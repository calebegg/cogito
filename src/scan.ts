import {error} from './main.ts';

export enum TokenType {
  // Single-character tokens
  LEFT_PAREN,
  RIGHT_PAREN,
  LEFT_BRACE,
  RIGHT_BRACE,
  LEFT_BRACKET,
  RIGHT_BRACKET,
  COMMA,
  DOT,
  DOT_DOT_DOT,
  MINUS,
  PLUS,
  SEMICOLON,
  COLON,
  SLASH,
  STAR,

  // One or two character tokens
  BANG,
  BANG_EQUAL,
  EQUAL,
  EQUAL_EQUAL,
  GREATER,
  GREATER_EQUAL,
  LESS,
  LESS_EQUAL,

  // Literals
  IDENTIFIER,
  STRING,
  NUMBER,

  // Keywords
  ASSERT,
  CONST,
  ELSE,
  FALSE,
  FUNCTION,
  IF,
  MAIN,
  NEW,
  NIL,
  PRINT,
  REDUCE,
  RETURN,
  STRUCT,
  THEOREM,
  TRUE,

  EOF,
}

const KEYWORDS = new Map([
  ['assert', TokenType.ASSERT],
  ['const', TokenType.CONST],
  ['else', TokenType.ELSE],
  ['false', TokenType.FALSE],
  ['function', TokenType.FUNCTION],
  ['if', TokenType.IF],
  ['main', TokenType.MAIN],
  ['new', TokenType.NEW],
  ['nil', TokenType.NIL],
  ['print', TokenType.PRINT],
  ['reduce', TokenType.REDUCE],
  ['return', TokenType.RETURN],
  ['struct', TokenType.STRUCT],
  ['theorem', TokenType.THEOREM],
  ['true', TokenType.TRUE],
]);

export interface Token {
  type: TokenType;
  lexeme: string;
  line: number;
  literal?: string;
}

export function scan(source: string): Token[] {
  const tokens: Token[] = [];
  let start = 0;
  let current = 0;
  let line = 1;

  function addToken(type: TokenType, literalMap = (x: string) => x) {
    tokens.push({
      type,
      lexeme: source.substring(start, current),
      line,
      literal: literalMap(source.substring(start, current)),
    });
  }

  while (current < source.length) {
    start = current;
    const c = source.charAt(current);
    current++;
    switch (c) {
      case '(':
        addToken(TokenType.LEFT_PAREN);
        break;
      case ')':
        addToken(TokenType.RIGHT_PAREN);
        break;
      case '{':
        addToken(TokenType.LEFT_BRACE);
        break;
      case '}':
        addToken(TokenType.RIGHT_BRACE);
        break;
      case '[':
        addToken(TokenType.LEFT_BRACE);
        break;
      case ']':
        addToken(TokenType.RIGHT_BRACE);
        break;
      case ',':
        addToken(TokenType.COMMA);
        break;
      case '-':
        addToken(TokenType.MINUS);
        break;
      case '+':
        addToken(TokenType.PLUS);
        break;
      case ';':
        addToken(TokenType.SEMICOLON);
        break;
      case ':':
        addToken(TokenType.COLON);
        break;
      case '.':
        if (source.charAt(current) === '.') {
          if (source.charAt(current + 1) !== '.') {
            throw error(
              line,
              "'.' is an operator and '...' is an operator but '..' is not a thing"
            );
          }
          current += 2;
          addToken(TokenType.DOT_DOT_DOT);
        } else {
          addToken(TokenType.DOT);
        }
        break;
      case '!':
        if (source.charAt(current) === '=') {
          addToken(TokenType.BANG_EQUAL);
          current++;
        } else {
          addToken(TokenType.BANG);
        }
        break;
      case '=':
        if (source.charAt(current) === '=') {
          current++;
          addToken(TokenType.EQUAL_EQUAL);
        } else {
          addToken(TokenType.EQUAL);
        }
        break;
      case '<':
        if (source.charAt(current) === '=') {
          current++;
          addToken(TokenType.LESS_EQUAL);
        } else {
          addToken(TokenType.LESS);
        }
        break;
      case '>':
        if (source.charAt(current) === '=') {
          current++;
          addToken(TokenType.GREATER_EQUAL);
        } else {
          addToken(TokenType.GREATER);
        }
        break;
      case '/':
        if (source.charAt(current) === '/') {
          while (source.charAt(current) !== '\n' && current < source.length) {
            current++;
          }
        } else {
          addToken(TokenType.SLASH);
        }
        break;
      case '"':
        while (source.charAt(current) !== '"' && current < source.length) {
          if (source.charAt(current) === '\n') {
            line++;
          }
          current++;
        }
        if (current >= source.length) {
          error(line, 'Unterminated string.');
        }
        current++;
        addToken(TokenType.STRING);
        break;
      case '|':
        while (source.charAt(current) !== '|' && current < source.length) {
          if (source.charAt(current) === '\n') {
            error(line, "Unexpected end of line while looking for '|'");
            break;
          }
          current++;
        }
        if (current >= source.length) {
          error(line, 'Unterminated identifier.');
        }
        current++;
        addToken(TokenType.IDENTIFIER);
        break;
      case ' ':
      case '\r':
      case '\t':
        // Ignore whitespace.
        break;
      case '\n':
        line++;
        break;
      default:
        if (c.match(/[0-9]/)) {
          while (source.charAt(current).match(/[0-9_]/)) {
            current++;
          }
          if (source.charAt(current) === '/') {
            current++;
            while (source.charAt(current).match(/[0-9_]/)) {
              current++;
            }
          }
          addToken(TokenType.NUMBER, n => n.replaceAll(/_/g, ''));
          continue;
        }
        if (c.match(/[a-zA-Z_*]/)) {
          if (c === '*' && !source.charAt(current).match(/[a-zA-Z0-9_-]/)) {
            addToken(TokenType.STAR);
            continue;
          }
          while (source.charAt(current).match(/[a-zA-Z0-9_*-]/)) {
            current++;
          }
          addToken(
            KEYWORDS.get(source.substring(start, current)) ??
              TokenType.IDENTIFIER
          );
          continue;
        }
        error(line, 'Unexpected character.');
        break;
    }
  }
  return [...tokens, {type: TokenType.EOF, lexeme: '', line}];
}
