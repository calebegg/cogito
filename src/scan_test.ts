/**
 * @license
 * Copyright 2024 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { assertSnapshot } from 'https://deno.land/std@0.224.0/testing/snapshot.ts';
import { scan } from './scan.ts';
import outdent from 'https://deno.land/x/outdent@v0.8.0/mod.ts';
import { assertThrows } from 'https://deno.land/std@0.207.0/assert/assert_throws.ts';

Deno.test('scan a basic function', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        function foo(x: list<number>) {
          return 1;
        }
      `,
    ),
  );
});

Deno.test('scan a basic theorem', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        theorem |foo works|(x: number) {
          return foo(x) > 0
        }
      `,
    ),
  );
});

Deno.test('scan a basic main', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        main {
          print("hello, world!");
        }
      `,
    ),
  );
});

Deno.test('scan a basic const', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        const *foo* = 1_000_000/1_000;
      `,
    ),
  );
});

Deno.test('scan a basic struct', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        struct foo(x: number, y: number);
      `,
    ),
  );
});

Deno.test('scan a basic assert', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        assert(x > 0);
      `,
    ),
  );
});

Deno.test('scan a basic print', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        print("hello,
        world!");
      `,
    ),
  );
});

Deno.test('scan a basic return', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        return 1;
      `,
    ),
  );
});

Deno.test('scan a basic assign', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        x = 1;
      `,
    ),
  );
});

Deno.test('scan some operators', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        == >= <= != < > = . * / + - **;
      `,
    ),
  );
});

Deno.test('scan a commnet', async (t) => {
  await assertSnapshot(
    t,
    scan(
      outdent`
        a = 1; // this is important!
      `,
    ),
  );
});

Deno.test('.. throws', () => {
  assertThrows(() => scan('..'), Error, 'not a thing');
});

Deno.test('# throws', () => {
  assertThrows(() => scan('#'), Error, 'Unexpected character');
});

Deno.test('Unterminated string throws', () => {
  assertThrows(() => scan('"foo'), Error, 'Unterminated string');
});

Deno.test('Unterminated ID throws', () => {
  assertThrows(() => scan('|foo'), Error, 'Unterminated identifier');
  assertThrows(() => scan('|foo\nfoo|'), Error, 'Unexpected end of line');
});
