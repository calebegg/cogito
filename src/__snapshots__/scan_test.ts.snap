export const snapshot = {};

snapshot[`scan a basic function 1`] = `
[
  {
    char: 1,
    lexeme: "function",
    line: 1,
    literal: "function",
    type: 38,
  },
  {
    char: 10,
    lexeme: "foo",
    line: 1,
    literal: "foo",
    type: 28,
  },
  {
    char: 13,
    lexeme: "(",
    line: 1,
    literal: "(",
    type: 14,
  },
  {
    char: 14,
    lexeme: "x",
    line: 1,
    literal: "x",
    type: 28,
  },
  {
    char: 15,
    lexeme: ":",
    line: 1,
    literal: ":",
    type: 3,
  },
  {
    char: 17,
    lexeme: "list<number>",
    line: 1,
    literal: "list<number>",
    type: 28,
  },
  {
    char: 29,
    lexeme: ")",
    line: 1,
    literal: ")",
    type: 22,
  },
  {
    char: 31,
    lexeme: "{",
    line: 1,
    literal: "{",
    type: 12,
  },
  {
    char: 3,
    lexeme: "return",
    line: 2,
    literal: "return",
    type: 47,
  },
  {
    char: 10,
    lexeme: "1",
    line: 2,
    literal: "1",
    type: 30,
  },
  {
    char: 11,
    lexeme: ";",
    line: 2,
    literal: ";",
    type: 23,
  },
  {
    char: 1,
    lexeme: "}",
    line: 3,
    literal: "}",
    type: 20,
  },
  {
    char: 0,
    lexeme: "",
    line: 3,
    type: 52,
  },
]
`;

snapshot[`scan a basic theorem 1`] = `
[
  {
    char: 1,
    lexeme: "theorem",
    line: 1,
    literal: "theorem",
    type: 50,
  },
  {
    char: 9,
    lexeme: "|foo works|",
    line: 1,
    literal: "|foo works|",
    type: 28,
  },
  {
    char: 20,
    lexeme: "(",
    line: 1,
    literal: "(",
    type: 14,
  },
  {
    char: 21,
    lexeme: "x",
    line: 1,
    literal: "x",
    type: 28,
  },
  {
    char: 22,
    lexeme: ":",
    line: 1,
    literal: ":",
    type: 3,
  },
  {
    char: 24,
    lexeme: "number",
    line: 1,
    literal: "number",
    type: 28,
  },
  {
    char: 30,
    lexeme: ")",
    line: 1,
    literal: ")",
    type: 22,
  },
  {
    char: 32,
    lexeme: "{",
    line: 1,
    literal: "{",
    type: 12,
  },
  {
    char: 3,
    lexeme: "return",
    line: 2,
    literal: "return",
    type: 47,
  },
  {
    char: 10,
    lexeme: "foo",
    line: 2,
    literal: "foo",
    type: 28,
  },
  {
    char: 13,
    lexeme: "(",
    line: 2,
    literal: "(",
    type: 14,
  },
  {
    char: 14,
    lexeme: "x",
    line: 2,
    literal: "x",
    type: 28,
  },
  {
    char: 15,
    lexeme: ")",
    line: 2,
    literal: ")",
    type: 22,
  },
  {
    char: 17,
    lexeme: ">",
    line: 2,
    literal: ">",
    type: 11,
  },
  {
    char: 19,
    lexeme: "0",
    line: 2,
    literal: "0",
    type: 30,
  },
  {
    char: 1,
    lexeme: "}",
    line: 3,
    literal: "}",
    type: 20,
  },
  {
    char: 0,
    lexeme: "",
    line: 3,
    type: 52,
  },
]
`;

snapshot[`scan a basic main 1`] = `
[
  {
    char: 1,
    lexeme: "main",
    line: 1,
    literal: "main",
    type: 41,
  },
  {
    char: 6,
    lexeme: "{",
    line: 1,
    literal: "{",
    type: 12,
  },
  {
    char: 3,
    lexeme: "print",
    line: 2,
    literal: "print",
    type: 46,
  },
  {
    char: 8,
    lexeme: "(",
    line: 2,
    literal: "(",
    type: 14,
  },
  {
    char: 9,
    lexeme: '"hello, world!"',
    line: 2,
    literal: '"hello, world!"',
    type: 29,
  },
  {
    char: 24,
    lexeme: ")",
    line: 2,
    literal: ")",
    type: 22,
  },
  {
    char: 25,
    lexeme: ";",
    line: 2,
    literal: ";",
    type: 23,
  },
  {
    char: 1,
    lexeme: "}",
    line: 3,
    literal: "}",
    type: 20,
  },
  {
    char: 0,
    lexeme: "",
    line: 3,
    type: 52,
  },
]
`;

snapshot[`scan a basic const 1`] = `
[
  {
    char: 1,
    lexeme: "const",
    line: 1,
    literal: "const",
    type: 33,
  },
  {
    char: 7,
    lexeme: "*foo*",
    line: 1,
    literal: "*foo*",
    type: 28,
  },
  {
    char: 13,
    lexeme: "=",
    line: 1,
    literal: "=",
    type: 8,
  },
  {
    char: 15,
    lexeme: "1_000_000/1_000",
    line: 1,
    literal: "1000000/1000",
    type: 30,
  },
  {
    char: 30,
    lexeme: ";",
    line: 1,
    literal: ";",
    type: 23,
  },
  {
    char: 29,
    lexeme: "",
    line: 1,
    type: 52,
  },
]
`;

snapshot[`scan a basic struct 1`] = `
[
  {
    char: 1,
    lexeme: "struct",
    line: 1,
    literal: "struct",
    type: 48,
  },
  {
    char: 8,
    lexeme: "foo",
    line: 1,
    literal: "foo",
    type: 28,
  },
  {
    char: 11,
    lexeme: "(",
    line: 1,
    literal: "(",
    type: 14,
  },
  {
    char: 12,
    lexeme: "x",
    line: 1,
    literal: "x",
    type: 28,
  },
  {
    char: 13,
    lexeme: ":",
    line: 1,
    literal: ":",
    type: 3,
  },
  {
    char: 15,
    lexeme: "number",
    line: 1,
    literal: "number",
    type: 28,
  },
  {
    char: 21,
    lexeme: ",",
    line: 1,
    literal: ",",
    type: 4,
  },
  {
    char: 23,
    lexeme: "y",
    line: 1,
    literal: "y",
    type: 28,
  },
  {
    char: 24,
    lexeme: ":",
    line: 1,
    literal: ":",
    type: 3,
  },
  {
    char: 26,
    lexeme: "number",
    line: 1,
    literal: "number",
    type: 28,
  },
  {
    char: 32,
    lexeme: ")",
    line: 1,
    literal: ")",
    type: 22,
  },
  {
    char: 33,
    lexeme: ";",
    line: 1,
    literal: ";",
    type: 23,
  },
  {
    char: 32,
    lexeme: "",
    line: 1,
    type: 52,
  },
]
`;

snapshot[`scan a basic assert 1`] = `
[
  {
    char: 1,
    lexeme: "assert",
    line: 1,
    literal: "assert",
    type: 31,
  },
  {
    char: 7,
    lexeme: "(",
    line: 1,
    literal: "(",
    type: 14,
  },
  {
    char: 8,
    lexeme: "x",
    line: 1,
    literal: "x",
    type: 28,
  },
  {
    char: 10,
    lexeme: ">",
    line: 1,
    literal: ">",
    type: 11,
  },
  {
    char: 12,
    lexeme: "0",
    line: 1,
    literal: "0",
    type: 30,
  },
  {
    char: 13,
    lexeme: ")",
    line: 1,
    literal: ")",
    type: 22,
  },
  {
    char: 14,
    lexeme: ";",
    line: 1,
    literal: ";",
    type: 23,
  },
  {
    char: 13,
    lexeme: "",
    line: 1,
    type: 52,
  },
]
`;

snapshot[`scan a basic print 1`] = `
[
  {
    char: 1,
    lexeme: "print",
    line: 1,
    literal: "print",
    type: 46,
  },
  {
    char: 6,
    lexeme: "(",
    line: 1,
    literal: "(",
    type: 14,
  },
  {
    char: 7,
    lexeme: '"hello,
world!"',
    line: 2,
    literal: '"hello,
world!"',
    type: 29,
  },
  {
    char: 22,
    lexeme: ")",
    line: 2,
    literal: ")",
    type: 22,
  },
  {
    char: 23,
    lexeme: ";",
    line: 2,
    literal: ";",
    type: 23,
  },
  {
    char: 22,
    lexeme: "",
    line: 2,
    type: 52,
  },
]
`;

snapshot[`scan a basic return 1`] = `
[
  {
    char: 1,
    lexeme: "return",
    line: 1,
    literal: "return",
    type: 47,
  },
  {
    char: 8,
    lexeme: "1",
    line: 1,
    literal: "1",
    type: 30,
  },
  {
    char: 9,
    lexeme: ";",
    line: 1,
    literal: ";",
    type: 23,
  },
  {
    char: 8,
    lexeme: "",
    line: 1,
    type: 52,
  },
]
`;

snapshot[`scan a basic assign 1`] = `
[
  {
    char: 1,
    lexeme: "x",
    line: 1,
    literal: "x",
    type: 28,
  },
  {
    char: 3,
    lexeme: "=",
    line: 1,
    literal: "=",
    type: 8,
  },
  {
    char: 5,
    lexeme: "1",
    line: 1,
    literal: "1",
    type: 30,
  },
  {
    char: 6,
    lexeme: ";",
    line: 1,
    literal: ";",
    type: 23,
  },
  {
    char: 5,
    lexeme: "",
    line: 1,
    type: 52,
  },
]
`;

snapshot[`scan some operators 1`] = `
[
  {
    char: 1,
    lexeme: "==",
    line: 1,
    literal: "==",
    type: 7,
  },
  {
    char: 4,
    lexeme: ">=",
    line: 1,
    literal: ">=",
    type: 10,
  },
  {
    char: 7,
    lexeme: "<=",
    line: 1,
    literal: "<=",
    type: 15,
  },
  {
    char: 10,
    lexeme: "!=",
    line: 1,
    literal: "!=",
    type: 1,
  },
  {
    char: 13,
    lexeme: "<",
    line: 1,
    literal: "<",
    type: 16,
  },
  {
    char: 15,
    lexeme: ">",
    line: 1,
    literal: ">",
    type: 11,
  },
  {
    char: 17,
    lexeme: "=",
    line: 1,
    literal: "=",
    type: 8,
  },
  {
    char: 19,
    lexeme: ".",
    line: 1,
    literal: ".",
    type: 6,
  },
  {
    char: 21,
    lexeme: "*",
    line: 1,
    literal: "*",
    type: 27,
  },
  {
    char: 23,
    lexeme: "/",
    line: 1,
    literal: "/",
    type: 25,
  },
  {
    char: 25,
    lexeme: "+",
    line: 1,
    literal: "+",
    type: 19,
  },
  {
    char: 27,
    lexeme: "-",
    line: 1,
    literal: "-",
    type: 17,
  },
  {
    char: 29,
    lexeme: "**",
    line: 1,
    literal: "**",
    type: 26,
  },
  {
    char: 31,
    lexeme: ";",
    line: 1,
    literal: ";",
    type: 23,
  },
  {
    char: 30,
    lexeme: "",
    line: 1,
    type: 52,
  },
]
`;

snapshot[`scan a commnet 1`] = `
[
  {
    char: 1,
    lexeme: "a",
    line: 1,
    literal: "a",
    type: 28,
  },
  {
    char: 3,
    lexeme: "=",
    line: 1,
    literal: "=",
    type: 8,
  },
  {
    char: 5,
    lexeme: "1",
    line: 1,
    literal: "1",
    type: 30,
  },
  {
    char: 6,
    lexeme: ";",
    line: 1,
    literal: ";",
    type: 23,
  },
  {
    char: 7,
    lexeme: "",
    line: 1,
    type: 52,
  },
]
`;
